contracts/p1/RToken.sol

```diff
@@ -178,79 +182,159 @@ contract RTokenP1 is ComponentP1, ERC20PermitUpgradeable, IRToken {
-    /// @custom:interaction
-    function redeemTo(
+    /// @param basketNonces An array of basket nonces to do redemption from
+    /// @param portions {1} An array of Fix quantities that must add up to FIX_ONE
+    /// @param erc20sOut An array of ERC20s expected out
+    /// @param minAmounts {qTok} The minimum ERC20 quantities the caller should receive
+    /// @custom:interaction RCEI
+    function redeemToCustom(
         address recipient,
         uint256 amount,
-        uint48 basketNonce
-    ) public notFrozen exchangeRateIsValidAfter {
+        uint48[] memory basketNonces,
+        uint192[] memory portions,
+        IERC20[] memory erc20sOut,
+        uint256[] memory minAmounts
+    ) external notFrozen exchangeRateIsValidAfter {
         // == Refresh ==
+
         assetRegistry.refresh();
+        main.furnace().melt();
 
         // == Checks and Effects ==
-        address redeemer = _msgSender();
+
         require(amount > 0, "Cannot redeem zero");
-        require(amount <= balanceOf(redeemer), "insufficient balance");
+        require(amount <= balanceOf(_msgSender()), "insufficient balance");
+        uint256 portionsSum;
+        for (uint256 i = 0; i < portions.length; ++i) {
+            portionsSum += portions[i];
+        }
+        require(portionsSum == FIX_ONE, "portions do not add up to FIX_ONE");
 
-        // Failure to melt results in a lower redemption price, so we can allow it when paused
-        // solhint-disable-next-line no-empty-blocks
-        try main.furnace().melt() {} catch {}
         uint256 supply = totalSupply();
 
+        // Accept and burn RToken, reverts if not enough balance to burn
+        _burn(_msgSender(), amount);
+
         // Revert if redemption exceeds either supply throttle
         issuanceThrottle.useAvailable(supply, -int256(amount));
         redemptionThrottle.useAvailable(supply, int256(amount)); // reverts on over-redemption
 
-        // ==== Get basket redemption ====
-        // i.e, set (erc20s, amounts) = basketHandler.quote(amount * basketsNeeded / totalSupply)
-
         // D18{BU} = D18{BU} * {qRTok} / {qRTok}
-        // downcast is safe: amount < totalSupply and basketsNeeded < 1e57 < 2^190 (just barely)
         uint192 basketsRedeemed = basketsNeeded.muluDivu(amount, supply); // FLOOR
-        emit Redemption(redeemer, recipient, amount, basketsRedeemed);
+        emit Redemption(_msgSender(), recipient, amount, basketsRedeemed);
+        emit BasketsNeededChanged(basketsNeeded, basketsNeeded - basketsRedeemed);
+        basketsNeeded = basketsNeeded - basketsRedeemed;
 
-        require(basketHandler.nonce() == basketNonce, "non-current basket nonce");
-        (address[] memory erc20s, uint256[] memory amounts) = basketHandler.quote(
-            basketsRedeemed,
-            FLOOR
+        // === Get basket redemption amounts ===
+
+        (address[] memory erc20s, uint256[] memory amounts) = basketHandler.quoteCustomRedemption(
```

Comment:
maybe rename to `quotedERC20s` and `quotedAmounts` for clarity

Reply:
I also think I want to return the amounts transferred, so this can be used in simulation easily.